### PR TITLE
Add error explanations for E0374, E0375, E0376 on issue #32777

### PR DESCRIPTION
**Pull request by dfockler**

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -3284,6 +3284,82 @@ impl Baz for Bar { } // Note: This is OK
 ```
 "##,
 
+E0374: r##"
+The trait `CoerceUnsized` may only be implemented for a coercion between
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: This introduction is confusing. A lot of repetitions make the whole harder than necessary to understand.

File: src/librustc_typeck/diagnostics.rs [hunk 2]
```diff
@@ -3284,6 +3284,82 @@ impl Baz for Bar { } // Note: This is OK
...
+structures with one field being coerced, none were found. The structure you
+are trying to implement `CoerceUnsized` for doesn't have any Unsized fields
+so it doesn't need CoerceUnsized.
+
+Example:
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Take a look at https://github.com/rust-lang/rfcs/pull/1567

File: src/librustc_typeck/diagnostics.rs [hunk 3]
```diff
@@ -3284,6 +3284,82 @@ impl Baz for Bar { } // Note: This is OK
...
+#![feature(coerce_unsized)]
+struct Foo<T> {
+    a: i32,
+}
+
+// error: Struct `Foo` has no unsized fields that need `CoerceUnsized`.
+impl<T, U> CoerceUnsized<Foo<U>> for Foo<T> {}
+```
+
+`CoerceUnsized` is used to coerce one unsized structure into another. If the
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: I'm wondering if an additional explanation about "unsized structure" wouldn't be necessary.

File: src/librustc_typeck/diagnostics.rs [hunk 4]
```diff
@@ -3284,6 +3284,82 @@ impl Baz for Bar { } // Note: This is OK
...
+and can use regular type coercion instead.
+"##,
+
+E0375: r##"
+The trait `CoerceUnsized` may only be implemented for a coercion between
+structures with one field being coerced, but multiple fields need coercions.
+The structure you are trying to implement `CoerceUnsized` for has more than
+one unsized field that would need to be coerced.
+
+Example:
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Same comment as above: take a look at https://github.com/rust-lang/rfcs/pull/1567 about "Example of erroneous code".

File: src/librustc_typeck/diagnostics.rs [hunk 5]
```diff
@@ -3284,6 +3284,82 @@ impl Baz for Bar { } // Note: This is OK
...
+
+`CoerceUnsized` is used to coerce one unsized structure into another. If the
+structure doesn't have any unsized fields, it's not an unsized structure
+and can use regular type coercion instead.
+"##,
+
+E0375: r##"
+The trait `CoerceUnsized` may only be implemented for a coercion between
+structures with one field being coerced, but multiple fields need coercions.
+The structure you are trying to implement `CoerceUnsized` for has more than
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Try to keep a neutral "way" to write introduction please.

File: src/librustc_typeck/diagnostics.rs [hunk 6]
```diff
@@ -3284,6 +3284,82 @@ impl Baz for Bar { } // Note: This is OK
...
+
+`CoerceUnsized` only allows for coercion from a structure with a single
+unsized field to another structure with a single unsized field.
+"##,
+
+E0376: r##"
+The trait `CoerceUnsized` may only be implemented for a coercion between
+structures. The structure you are trying to coerce may not be a structure.
+
+Example:
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Same as the two above: https://github.com/rust-lang/rfcs/pull/1567

File: src/librustc_typeck/diagnostics.rs [hunk 7]
```diff
@@ -3284,6 +3284,82 @@ impl Baz for Bar { } // Note: This is OK
...
+struct Foo<T> {
+    a: T,
+}
+
+// error: The type `U` is not a structure
+impl<T, U> CoerceUnsized<U> for Foo<T> {}
+```
+
+The `CoerceUnsized` trait takes a structure. Make sure the type you are
+providing to `CoerceUnsized` is an unsized structure.
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Missing "Example:" with code annotations.

File: src/librustc_typeck/diagnostics.rs [hunk 8]
```diff
@@ -3777,12 +3853,12 @@ register_diagnostics! {
     E0320, // recursive overflow during dropck
     E0328, // cannot implement Unsize explicitly
 //  E0372, // coherence not object safe
-    E0374, // the trait `CoerceUnsized` may only be implemented for a coercion
+//  E0374, // the trait `CoerceUnsized` may only be implemented for a coercion
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Please remove them, don't comment them out.

File: src/librustc_typeck/diagnostics.rs [hunk 9]
```diff
@@ -3777,12 +3853,12 @@ register_diagnostics! {
...
            // between structures with one field being coerced, none found
-    E0375, // the trait `CoerceUnsized` may only be implemented for a coercion
+//  E0375, // the trait `CoerceUnsized` may only be implemented for a coercion
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Same here.

File: src/librustc_typeck/diagnostics.rs [hunk 10]
```diff
@@ -3777,12 +3853,12 @@ register_diagnostics! {
...
            // between structures with one field being coerced, but multiple
            // fields need coercions
-    E0376, // the trait `CoerceUnsized` may only be implemented for a coercion
+//  E0376, // the trait `CoerceUnsized` may only be implemented for a coercion
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Same here.

**Comment by GuillaumeGomez** [comment]
Thanks for your work! Globally, your explanations aren't clear enough. Also, if possible, add examples that show how to fix the error (with code annotation to make it even better).

**Comment by GuillaumeGomez** [comment]
cc #32777

**Comment by dfockler** [comment]
@GuillaumeGomez Thanks for the feedback! I'll read through that RFC more closely and get some changes up soon.

**Comment by GuillaumeGomez** [comment]
Thanks! Don't forget to ping me once done (github doesn't notify when PR is updated).

File: src/librustc_typeck/diagnostics.rs (hunk 5 above)
Comment (pnkfelix) [inline src/librustc_typeck/diagnostics.rs]: @GuillaumeGomez sorry I'm a little dense (or perhaps insensitve): can you clarify why the above introduction is not neutral?

**Comment by pnkfelix** [comment]
(I don't really have anything to add to the review at this point. I'll wait to see how @dfockler responds to @GuillaumeGomez 's feedback before r+'ing this.)

File: src/librustc_typeck/diagnostics.rs (hunk 1 above)
Comment (dfockler) [inline src/librustc_typeck/diagnostics.rs]: Oh sorry. Just a note, I was thinking that you were supposed to put the actual error text before the error description.

File: src/librustc_typeck/diagnostics.rs (hunk 5 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: He certainly rewrote this part.

**Comment by dfockler** [comment]
@GuillaumeGomez I rewrote these to be comply with the rfc. If these are good let me know if I should squash them or not, Thanks.

File: src/librustc_typeck/diagnostics.rs [hunk 11]
```diff
@@ -3284,6 +3284,167 @@ impl Baz for Bar { } // Note: This is OK
 ```
 "##,
 
+E0374: r##"
+The struct you are trying to implement `CoerceUnsized` for doesn't have
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: This still isn't a neutral wording.

File: src/librustc_typeck/diagnostics.rs [hunk 12]
```diff
@@ -3284,6 +3284,167 @@ impl Baz for Bar { } // Note: This is OK
...
+    where T: CoerceUnsized<U> {}
+```
+
+Note that `CoerceUnsized` is mainly used by smart pointers like `Box`, `Rc`
+and `Arc` to be able to mark that they can coerce unsized types that they
+are pointing at.
+"##,
+
+E0375: r##"
+The struct you are trying to implement `CoerceUnsized` for has more than
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Not a neutral wording as well.

File: src/librustc_typeck/diagnostics.rs [hunk 13]
```diff
@@ -3284,6 +3284,167 @@ impl Baz for Bar { } // Note: This is OK
...
+The type you are trying to impl `CoerceUnsized` for is not a struct.
+`CoerceUnsized` can only be implemented for a struct. Unsized types are
+already able to be coerced without an implementation of `CoerceUnsized`
+whereas a struct containing an unsized type needs to know the unsized type
+field it's containing is able to be coerced. An 
+[unsized type](https://doc.rust-lang.org/book/unsized-types.html)
+is any type that the compiler doesn't know the length or alignment of at 
+compile time. Any struct containing an unsized type is also unsized.  
+
+
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Extra empty line.

File: src/librustc_typeck/diagnostics.rs [hunk 14]
```diff
@@ -3284,6 +3284,167 @@ impl Baz for Bar { } // Note: This is OK
...
+struct Foo<T> {
+    a: T,
+}
+
+// The `Foo<U>` is a struct so `CoerceUnsized` can be implemented 
+impl<T, U> CoerceUnsized<Foo<U>> for Foo<T> where T: CoerceUnsized<U> {}
+```
+
+Note that in Rust structs can only contain an unsized type if the field
```
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: I think a comma is missing in this sentence.

**Comment by GuillaumeGomez** [comment]
> @GuillaumeGomez I rewrote these to comply with the rfc. If these are good let me know if I should squash them or not, Thanks.

At the end, we'll ask you to squash your commits if you haven't already done it. But meantime, it's entirely up to you.

A few nits to fix and I think we'll be good.

**Comment by dfockler** [comment]
@GuillaumeGomez Ok, those issues are fixed.

**Comment by GuillaumeGomez** [comment]
Travis says:

```Shell
/home/travis/build/rust-lang/rust/src/librustc_typeck/diagnostics.rs:3289: trailing whitespace
/home/travis/build/rust-lang/rust/src/librustc_typeck/diagnostics.rs:3291: trailing whitespace
/home/travis/build/rust-lang/rust/src/librustc_typeck/diagnostics.rs:3346: trailing whitespace
/home/travis/build/rust-lang/rust/src/librustc_typeck/diagnostics.rs:3350: trailing whitespace
/home/travis/build/rust-lang/rust/src/librustc_typeck/diagnostics.rs:3351: trailing whitespace
/home/travis/build/rust-lang/rust/src/librustc_typeck/diagnostics.rs:3392: trailing whitespace
/home/travis/build/rust-lang/rust/src/librustc_typeck/diagnostics.rs:3403: trailing whitespace
/home/travis/build/rust-lang/rust/src/librustc_typeck/diagnostics.rs:3405: trailing whitespace
/home/travis/build/rust-lang/rust/src/librustc_typeck/diagnostics.rs:3406: trailing whitespace
/home/travis/build/rust-lang/rust/src/librustc_typeck/diagnostics.rs:3436: trailing whitespace
```

To check such lints, run "make tidy".

**Comment by dfockler** [comment]
@GuillaumeGomez  Ok, whitespace fixed.

**Comment by dfockler** [comment]
@GuillaumeGomez I squashed the commits and it passed on the last check.

**Comment by GuillaumeGomez** [comment]
Great, thanks!

cc @Manishearth, @steveklabnik

**Pull request merged**